binderhub/redirect.py

```diff
@@ -76,14 +76,33 @@ def username_from_image(self, image):
         # start server
         app_log.info("Starting server for user %s with image %s", username, image)
         try:
-            await self.api_request(
+            resp = await self.api_request(
                 'users/%s/server' % username,
                 method='POST',
                 body=json.dumps({
                     'token': token,
                     'image': image,
                 }).encode('utf8'),
             )
+            if resp.code == 202:
+                # Server hasn't actually started yet
+                # We wait for it!
+                for i in range(10):
+                    resp = await self.api_request(
+                        'users/%s' % username,
+                        method='GET',
+                    )
+
+                    body = json.loads(resp.body.decode('utf-8'))
+                    if body['server']:
+                        break
+                    # FIXME: make this configurable
+                    # FIXME: Measure how long it takes for servers to start
+                    # and tune this appropriately
+                    await gen.sleep(1.4 ** i)
+                else:
+                    raise web.HTTPError(500, "Image %s for user %s took too long to launch" % (image, username))
```

Comment:
the chart config could be to set a max wait time, and then this could just split that time up into logarithmically-sized chunks (so earlier chunks are shorter than later ones)

Reply:
It is! We have a utility function in jupyterhub to do that for us (along with some jitter). Might copy and use that here.

Reply:
+1 to using the jitter utility.